pkg/sti/executor.go

```diff
@@ -119,23 +130,121 @@ func (h *requestHandler) execute(command string) error {
 	}
 	defer tarFile.Close()
 
+	expectedError := ""
+	outReader, outWriter := io.Pipe()
+	errReader, errWriter := io.Pipe()
+	defer outReader.Close()
+	defer outWriter.Close()
+	defer errReader.Close()
+	defer errWriter.Close()
 	opts := docker.RunContainerOptions{
-		Image:     h.request.BaseImage,
-		Stdin:     tarFile,
-		Stdout:    os.Stdout,
-		Stderr:    os.Stderr,
-		PullImage: true,
-		Command:   command,
-		Env:       h.generateConfigEnv(),
-		PostExec:  h,
-	}
-	return h.docker.RunContainer(opts)
+		Image:           h.request.BaseImage,
+		Stdout:          outWriter,
+		Stderr:          errWriter,
+		PullImage:       h.request.ForcePull,
+		ExternalScripts: h.request.ExternalRequiredScripts,
+		ScriptsURL:      h.request.ScriptsURL,
+		Location:        h.request.Location,
+		Command:         command,
+		Env:             h.generateConfigEnv(),
+		PostExec:        h,
+	}
+	if !h.request.LayeredBuild {
+		opts.Stdin = tarFile
+	}
+	// goroutine to stream container's output
+	go func(reader io.Reader) {
+		scanner := bufio.NewScanner(reader)
+		for scanner.Scan() {
+			if glog.V(2) || command == api.Usage {
+				glog.Info(scanner.Text())
+			}
+		}
+	}(outReader)
+	// goroutine to stream container's error
+	go func(reader io.Reader) {
+		scanner := bufio.NewScanner(reader)
+		for scanner.Scan() {
+			text := scanner.Text()
+			if glog.V(1) {
+				glog.Errorf(text)
+			}
+			if h.errorChecker != nil && h.errorChecker.wasExpectedError(text) {
+				expectedError += text + "; "
```

Comment:
Got it... the concern was that we would create a really large string if we have a very long stderr output. Maybe we then limit it by length? (ie. if len(expectedError) < SizeLimit && h.errorChecker != nil && ... )